library/Booster/Pattern/Rewrite.hs

```diff
@@ -128,17 +131,101 @@ rewriteStep cutLabels terminalLabels pat = do
             uniqueId = (.uniqueId) . (.attributes)
 
         case results of
+            -- no rules in this group were applicable
             [] ->
                 processGroups rest
-            [(r, x)]
+            -- there was a single rule which applied
+            [Applied (r, x)]
                 | labelOf r `elem` cutLabels ->
                     pure $ RewriteCutPoint (labelOf r) (uniqueId r) pat x
                 | labelOf r `elem` terminalLabels ->
                     pure $ RewriteTerminal (labelOf r) (uniqueId r) x
                 | otherwise ->
                     pure $ RewriteFinished (Just $ ruleLabelOrLocT r) (uniqueId r) x
-            rxs ->
-                pure $ RewriteBranch pat $ NE.fromList $ map (\(r, p) -> (labelOf r, uniqueId r, p)) rxs
+            rxs
+                -- all rules which did apply had an ensures condition which evaluated to false
+                | all (== Trivial) rxs ->
+                    -- if, all the other groups only generate a not applicable or trivial rewrites,
+                    -- then we return a `RewriteTrivial`. To ensure this, we need to catch the
+                    -- `NoApplicableRules` exception which is thrown when we have exhausted all the rules.
+                    -- When this error is thrown, we know that all the following rule groups had only
+                    -- non-applicable rules.
+                    catchNoApplicableRulesWhenTrivial (processGroups rest) >>= \case
+                        RewriteTrivial{} -> pure $ RewriteTrivial pat
+                        other -> pure other
```

Comment:
I don't really like catching exceptions here... maybe  `processGroups` could signal the condition as a `RewriteResult` ? We cannot directly return `RewriteTrivial`, though, as we would need to remember there was already a trivial result. So maybe `NoApplicableRules` should be `RewriteResult` instead of `RewriteFailed`?